File: rplugin/python3/gdb/efmmgr.py
"""."""

from gdb.common import Common


class EfmMgr(Common):
    """Manager for the 'errorformat'."""

    def __init__(self, common):
        """Constructor."""
        super().__init__(common)
        self.counters = {}  # {efm: count}

    def cleanup(self):
        """Destructor."""
        for f in self.counters.keys():
            self.vim.command(f"set efm-={f}")

    def setup(self, formats):
        """Add 'efm' for some backend."""
        for f in formats:
            try:
                self.counters[f] += 1
            except KeyError:
                self.counters[f] = 1
                self.vim.command(f"set efm+={f}")

    def teardown(self, formats):
        """Remove 'efm' entries for some backend."""
        for f in formats:
            self.counters[f] -= 1
            if self.counters[f] < 1:
                del self.counters[f]
                self.vim.command(f"set efm-={f}")

File: lua/nvimgdb/efmmgr.lua
-- Manager for the 'errorformat'.
-- vim: set et ts=2 sw=2:

---@class EfmMgr errorformat manager
---@field private prev_o_efm string previous 'errorformat' value
local efmmgr = {
  prev_o_efm = '',
  prev_bo_efm = '',
}

---Add 'efm' for some backend.
---@param formats string[]
function efmmgr.setup(formats)
  efmmgr.prev_o_efm = vim.o.efm
  efmmgr.prev_bo_efm = vim.bo.efm
  vim.o.efm = ''
  for _, f in ipairs(formats) do
    vim.api.nvim_command("set efm+=" .. f)
  end
end

---Remove 'efm' entries for some backend.
function efmmgr.teardown()
  vim.o.efmmgr = efmmgr.prev_o_efm
  vim.bo.efmmgr = efmmgr.prev_bo_efm
end

return efmmgr
